c_src/sasl_auth.c

```diff
@@ -32,17 +33,86 @@ ERL_NIF_TERM ATOM_UNKNOWN;
 typedef struct {
     sasl_conn_t* conn;
     sasl_callback_t callbacks[16];
+    unsigned char* principal;
+    unsigned char* service;
+    unsigned char* host;
+    ErlNifPid controlling_process;
+    ErlNifMutex* controller_lock;
     int mech_set;
 } sasl_state_t;
 
 // SASL connection state
 static ErlNifResourceType* sasl_resource = NULL;
 
-static void destroy_resource(ErlNifEnv* UNUSED(env), sasl_state_t* res)
+static void destroy_resource(ErlNifEnv* UNUSED(env), sasl_state_t* state)
 {
-    if (res != NULL) {
-        sasl_dispose(&res->conn);
+    if (state != NULL) {
+        if (state->conn != NULL) {
+            if (state->controller_lock != NULL) {
+                enif_mutex_lock(state->controller_lock);
+            }
+
+            sasl_dispose(&state->conn);
+
+            if (state->controller_lock != NULL) {
+                enif_mutex_unlock(state->controller_lock);
+            }
+        }
+
+        if (state->controller_lock != NULL) {
+            enif_mutex_destroy(state->controller_lock);
+        }
+
+        if (state->principal != NULL) {
+            enif_free(state->principal);
+        }
+
+        if (state->service != NULL) {
+            enif_free(state->service);
+        }
+
+        if (state->host != NULL) {
+            enif_free(state->host);
+        }
+    }
+}
+
+static void* sasl_mutex_alloc(void) { return enif_mutex_create("sasl_auth.callback"); }
```

Comment:
The usage of mutexes here is simply per cyrus-sasl's own documentation which states (paraphrasing) "This library is **believed** to be thread safe if sasl mutexes are provided."